File: apps/dsa-web/src/types/portfolio.ts
export type PortfolioCostMethod = 'fifo' | 'avg';
export type PortfolioSide = 'buy' | 'sell';
export type PortfolioCashDirection = 'in' | 'out';
export type PortfolioCorporateActionType = 'cash_dividend' | 'split_adjustment';

export interface PortfolioAccountItem {
  id: number;
  ownerId?: string | null;
  name: string;
  broker?: string | null;
  market: 'cn' | 'hk' | 'us';
  baseCurrency: string;
  isActive: boolean;
  createdAt?: string | null;
  updatedAt?: string | null;
}

export interface PortfolioAccountListResponse {
  accounts: PortfolioAccountItem[];
}

export interface PortfolioAccountCreateRequest {
  name: string;
  broker?: string;
  market: 'cn' | 'hk' | 'us';
  baseCurrency: string;
  ownerId?: string;
}

export interface PortfolioPositionItem {
  symbol: string;
  market: string;
  currency: string;
  quantity: number;
  avgCost: number;
  totalCost: number;
  lastPrice: number;
  marketValueBase: number;
  unrealizedPnlBase: number;
  unrealizedPnlPct?: number | null;
  valuationCurrency: string;
  priceSource?: 'realtime_quote' | 'history_close' | 'missing' | string;
  priceProvider?: string | null;
  priceDate?: string | null;
  priceStale?: boolean;
  priceAvailable?: boolean;
}

export interface PortfolioAccountSnapshot {
  accountId: number;
  accountName: string;
  ownerId?: string | null;
  broker?: string | null;
  market: string;
  baseCurrency: string;
  asOf: string;
  costMethod: PortfolioCostMethod;
  totalCash: number;
  totalMarketValue: number;
  totalEquity: number;
  realizedPnl: number;
  unrealizedPnl: number;
  feeTotal: number;
  taxTotal: number;
  fxStale: boolean;
  positions: PortfolioPositionItem[];
}

export interface PortfolioSnapshotResponse {
  asOf: string;
  costMethod: PortfolioCostMethod;
  currency: string;
  accountCount: number;
  totalCash: number;
  totalMarketValue: number;
  totalEquity: number;
  realizedPnl: number;
  unrealizedPnl: number;
  feeTotal: number;
  taxTotal: number;
  fxStale: boolean;
  accounts: PortfolioAccountSnapshot[];
}

export interface PortfolioConcentrationItem {
  symbol: string;
  marketValueBase: number;
  weightPct: number;
  isAlert: boolean;
}

export interface PortfolioSectorConcentrationItem {
  sector: string;
  marketValueBase: number;
  weightPct: number;
  symbolCount: number;
  isAlert: boolean;
}

export interface PortfolioDrawdownBlock {
  seriesPoints: number;
  maxDrawdownPct: number;
  currentDrawdownPct: number;
  alert: boolean;
  fxStale: boolean;
}

export interface PortfolioStopLossItem {
  accountId: number;
  symbol: string;
  avgCost: number;
  lastPrice: number;
  lossPct: number;
  nearThresholdPct: number;
  isTriggered: boolean;
}

export interface PortfolioRiskResponse {
  asOf: string;
  accountId?: number | null;
  costMethod: PortfolioCostMethod;
  currency: string;
  thresholds: Record<string, number>;
  concentration: {
    totalMarketValue: number;
    topWeightPct: number;
    alert: boolean;
    topPositions: PortfolioConcentrationItem[];
  };
  sectorConcentration: {
    totalMarketValue: number;
    topWeightPct: number;
    alert: boolean;
    topSectors: PortfolioSectorConcentrationItem[];
    coverage: Record<string, number>;
    errors: string[];
  };
  drawdown: PortfolioDrawdownBlock;
  stopLoss: {
    nearAlert: boolean;
    triggeredCount: number;
    nearCount: number;
    items: PortfolioStopLossItem[];
  };
}

export interface PortfolioTradeCreateRequest {
  accountId: number;
  symbol: string;
  tradeDate: string;
  side: PortfolioSide;
  quantity: number;
  price: number;
  fee?: number;
  tax?: number;
  market?: 'cn' | 'hk' | 'us';
  currency?: string;
  tradeUid?: string;
  note?: string;
}

export interface PortfolioCashLedgerCreateRequest {
  accountId: number;
  eventDate: string;
  direction: PortfolioCashDirection;
  amount: number;
  currency?: string;
  note?: string;
}

export interface PortfolioCorporateActionCreateRequest {
  accountId: number;
  symbol: string;
  effectiveDate: string;
  actionType: PortfolioCorporateActionType;
  market?: 'cn' | 'hk' | 'us';
  currency?: string;
  cashDividendPerShare?: number;
  splitRatio?: number;
  note?: string;
}

export interface PortfolioEventCreatedResponse {
  id: number;
}

export interface PortfolioDeleteResponse {
  deleted: number;
}

export interface PortfolioTradeListItem {
  id: number;
  accountId: number;
  tradeUid?: string | null;
  symbol: string;
  market: string;
  currency: string;
  tradeDate: string;
  side: PortfolioSide;
  quantity: number;
  price: number;
  fee: number;
  tax: number;
  note?: string | null;
  createdAt?: string | null;
}

export interface PortfolioTradeListResponse {
  items: PortfolioTradeListItem[];
  total: number;
  page: number;
  pageSize: number;
}

export interface PortfolioCashLedgerListItem {
  id: number;
  accountId: number;
  eventDate: string;
  direction: PortfolioCashDirection;
  amount: number;
  currency: string;
  note?: string | null;
  createdAt?: string | null;
}

export interface PortfolioCashLedgerListResponse {
  items: PortfolioCashLedgerListItem[];
  total: number;
  page: number;
  pageSize: number;
}

export interface PortfolioCorporateActionListItem {
  id: number;
  accountId: number;
  symbol: string;
  market: string;
  currency: string;
  effectiveDate: string;
  actionType: PortfolioCorporateActionType;
  cashDividendPerShare?: number | null;
  splitRatio?: number | null;
  note?: string | null;
  createdAt?: string | null;
}

export interface PortfolioCorporateActionListResponse {
  items: PortfolioCorporateActionListItem[];
  total: number;
  page: number;
  pageSize: number;
}

export interface PortfolioImportTradeItem {
  tradeDate: string;
  symbol: string;
  side: PortfolioSide;
  quantity: number;
  price: number;
  fee: number;
  tax: number;
  tradeUid?: string | null;
  dedupHash: string;
  currency?: string | null;
}

export interface PortfolioImportParseResponse {
  broker: string;
  recordCount: number;
  skippedCount: number;
  errorCount: number;
  records: PortfolioImportTradeItem[];
  errors: string[];
}

export interface PortfolioImportCommitResponse {
  accountId: number;
  recordCount: number;
  insertedCount: number;
  duplicateCount: number;
  failedCount: number;
  dryRun: boolean;
  errors: string[];
}

export interface PortfolioImportBrokerItem {
  broker: string;
  aliases: string[];
  displayName?: string;
}

export interface PortfolioImportBrokerListResponse {
  brokers: PortfolioImportBrokerItem[];
}

export interface PortfolioFxRefreshResponse {
  asOf: string;
  accountCount: number;
  refreshEnabled?: boolean;
  disabledReason?: string | null;
  pairCount: number;
  updatedCount: number;
  staleCount: number;
  errorCount: number;
}

export interface AllocationTargetItem {
  id: number;
  accountId?: number | null;
  symbol?: string | null;
  sector?: string | null;
  targetPct: number;
  driftThresholdPct: number;
  priority: number;
  actualPct?: number | null;
  driftPct?: number | null;
  isAlert?: boolean | null;
}

export interface AllocationTargetCreateRequest {
  accountId?: number | null;
  symbol?: string | null;
  sector?: string | null;
  targetPct: number;
  driftThresholdPct?: number;
  priority?: number;
}

export interface AllocationTargetUpdateRequest {
  targetPct?: number | null;
  driftThresholdPct?: number | null;
  priority?: number | null;
}

export interface AllocationTargetListResponse {
  targets: AllocationTargetItem[];
}

export interface StagedRuleItem {
  id: number;
  accountId?: number | null;
  symbol: string;
  totalTargetShares: number;
  stageCount: number;
  stagePct1: number;
  stagePct2?: number | null;
  stagePct3?: number | null;
  stagePct4?: number | null;
  dipThresholdPct2?: number | null;
  dipThresholdPct3?: number | null;
  dipThresholdPct4?: number | null;
}

export interface StagedRuleCreateRequest {
  accountId?: number | null;
  symbol: string;
  totalTargetShares: number;
  stageCount?: number;
  stagePct1?: number;
  stagePct2?: number | null;
  stagePct3?: number | null;
  stagePct4?: number | null;
  dipThresholdPct2?: number | null;
  dipThresholdPct3?: number | null;
  dipThresholdPct4?: number | null;
}

export interface StagedRuleListResponse {
  rules: StagedRuleItem[];
}

export interface StagedBuyStageItem {
  stage: number;
  qty: number;
  priceCondition: string;
  triggerPrice: number;
}

export interface RebalanceSuggestionItem {
  id?: number | null;
  accountId?: number | null;
  suggestionType: string;
  symbol?: string | null;
  action: 'buy' | 'sell';
  quantity: number;
  estimatedPrice: number;
  estimatedAmount: number;
  reason?: string | null;
  allocationBeforePct: number;
  allocationAfterPct: number;
}

export interface StagedBuyItem {
  id: number;
  accountId?: number | null;
  symbol: string;
  currentPrice: number;
  currentShares: number;
  totalTargetShares: number;
  remainingToBuy: number;
  stages: StagedBuyStageItem[];
}

export interface CashReserveStatus {
  targetPct: number;
  actualPct: number;
  targetAmount: number;
  actualAmount: number;
  shortfallPct: number;
  hasShortfall: boolean;
}

export interface RebalanceReportResponse {
  asOf: string;
  accountId?: number | null;
  costMethod: string;
  hasTargets: boolean;
  allocationTargets: AllocationTargetItem[];
  driftAlertCount: number;
  suggestions: RebalanceSuggestionItem[];
  stagedBuys: StagedBuyItem[];
  cashReserve: CashReserveStatus;
}

File: api/v1/schemas/portfolio.py
# -*- coding: utf-8 -*-
"""Portfolio API schemas."""

from __future__ import annotations

from datetime import date
from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, Field


class PortfolioAccountCreateRequest(BaseModel):
    name: str = Field(..., min_length=1, max_length=64)
    broker: Optional[str] = Field(None, max_length=64)
    market: Literal["cn", "hk", "us"] = "cn"
    base_currency: str = Field("CNY", min_length=3, max_length=8)
    owner_id: Optional[str] = Field(None, max_length=64)


class PortfolioAccountUpdateRequest(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=64)
    broker: Optional[str] = Field(None, max_length=64)
    market: Optional[Literal["cn", "hk", "us"]] = None
    base_currency: Optional[str] = Field(None, min_length=3, max_length=8)
    owner_id: Optional[str] = Field(None, max_length=64)
    is_active: Optional[bool] = None


class PortfolioAccountItem(BaseModel):
    id: int
    owner_id: Optional[str] = None
    name: str
    broker: Optional[str] = None
    market: str
    base_currency: str
    is_active: bool
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class PortfolioAccountListResponse(BaseModel):
    accounts: List[PortfolioAccountItem] = Field(default_factory=list)


class PortfolioTradeCreateRequest(BaseModel):
    account_id: int
    symbol: str = Field(..., min_length=1, max_length=16)
    trade_date: date
    side: Literal["buy", "sell"]
    quantity: float = Field(..., gt=0)
    price: float = Field(..., gt=0)
    fee: float = Field(0.0, ge=0)
    tax: float = Field(0.0, ge=0)
    market: Optional[Literal["cn", "hk", "us"]] = None
    currency: Optional[str] = Field(None, min_length=3, max_length=8)
    trade_uid: Optional[str] = Field(None, max_length=128)
    note: Optional[str] = Field(None, max_length=255)


class PortfolioCashLedgerCreateRequest(BaseModel):
    account_id: int
    event_date: date
    direction: Literal["in", "out"]
    amount: float = Field(..., gt=0)
    currency: Optional[str] = Field(None, min_length=3, max_length=8)
    note: Optional[str] = Field(None, max_length=255)


class PortfolioCorporateActionCreateRequest(BaseModel):
    account_id: int
    symbol: str = Field(..., min_length=1, max_length=16)
    effective_date: date
    action_type: Literal["cash_dividend", "split_adjustment"]
    market: Optional[Literal["cn", "hk", "us"]] = None
    currency: Optional[str] = Field(None, min_length=3, max_length=8)
    cash_dividend_per_share: Optional[float] = Field(None, ge=0)
    split_ratio: Optional[float] = Field(None, gt=0)
    note: Optional[str] = Field(None, max_length=255)


class PortfolioEventCreatedResponse(BaseModel):
    id: int


class PortfolioDeleteResponse(BaseModel):
    deleted: int


class PortfolioTradeListItem(BaseModel):
    id: int
    account_id: int
    trade_uid: Optional[str] = None
    symbol: str
    market: str
    currency: str
    trade_date: str
    side: str
    quantity: float
    price: float
    fee: float
    tax: float
    note: Optional[str] = None
    created_at: Optional[str] = None


class PortfolioTradeListResponse(BaseModel):
    items: List[PortfolioTradeListItem] = Field(default_factory=list)
    total: int
    page: int
    page_size: int


class PortfolioCashLedgerListItem(BaseModel):
    id: int
    account_id: int
    event_date: str
    direction: str
    amount: float
    currency: str
    note: Optional[str] = None
    created_at: Optional[str] = None


class PortfolioCashLedgerListResponse(BaseModel):
    items: List[PortfolioCashLedgerListItem] = Field(default_factory=list)
    total: int
    page: int
    page_size: int


class PortfolioCorporateActionListItem(BaseModel):
    id: int
    account_id: int
    symbol: str
    market: str
    currency: str
    effective_date: str
    action_type: str
    cash_dividend_per_share: Optional[float] = None
    split_ratio: Optional[float] = None
    note: Optional[str] = None
    created_at: Optional[str] = None


class PortfolioCorporateActionListResponse(BaseModel):
    items: List[PortfolioCorporateActionListItem] = Field(default_factory=list)
    total: int
    page: int
    page_size: int


class PortfolioPositionItem(BaseModel):
    symbol: str
    market: str
    currency: str
    quantity: float
    avg_cost: float
    total_cost: float
    last_price: float
    market_value_base: float
    unrealized_pnl_base: float
    unrealized_pnl_pct: Optional[float] = None
    valuation_currency: str
    price_source: str = "unknown"
    price_provider: Optional[str] = None
    price_date: Optional[str] = None
    price_stale: bool = False
    price_available: bool = True


class PortfolioAccountSnapshot(BaseModel):
    account_id: int
    account_name: str
    owner_id: Optional[str] = None
    broker: Optional[str] = None
    market: str
    base_currency: str
    as_of: str
    cost_method: str
    total_cash: float
    total_market_value: float
    total_equity: float
    realized_pnl: float
    unrealized_pnl: float
    fee_total: float
    tax_total: float
    fx_stale: bool
    positions: List[PortfolioPositionItem] = Field(default_factory=list)


class PortfolioSnapshotResponse(BaseModel):
    as_of: str
    cost_method: str
    currency: str
    account_count: int
    total_cash: float
    total_market_value: float
    total_equity: float
    realized_pnl: float
    unrealized_pnl: float
    fee_total: float
    tax_total: float
    fx_stale: bool
    accounts: List[PortfolioAccountSnapshot] = Field(default_factory=list)


class PortfolioImportTradeItem(BaseModel):
    trade_date: str
    symbol: str
    side: Literal["buy", "sell"]
    quantity: float
    price: float
    fee: float
    tax: float
    trade_uid: Optional[str] = None
    dedup_hash: str
    currency: Optional[str] = None


class PortfolioImportParseResponse(BaseModel):
    broker: str
    record_count: int
    skipped_count: int
    error_count: int
    records: List[PortfolioImportTradeItem] = Field(default_factory=list)
    errors: List[str] = Field(default_factory=list)


class PortfolioImportCommitResponse(BaseModel):
    account_id: int
    record_count: int
    inserted_count: int
    duplicate_count: int
    failed_count: int
    dry_run: bool
    errors: List[str] = Field(default_factory=list)


class PortfolioImportBrokerItem(BaseModel):
    broker: str
    aliases: List[str] = Field(default_factory=list)
    display_name: Optional[str] = None


class PortfolioImportBrokerListResponse(BaseModel):
    brokers: List[PortfolioImportBrokerItem] = Field(default_factory=list)


class PortfolioFxRefreshResponse(BaseModel):
    as_of: str
    account_count: int
    refresh_enabled: bool
    disabled_reason: Optional[str] = None
    pair_count: int
    updated_count: int
    stale_count: int
    error_count: int


class PortfolioRiskResponse(BaseModel):
    as_of: str
    account_id: Optional[int] = None
    cost_method: str
    currency: str
    thresholds: Dict[str, Any] = Field(default_factory=dict)
    concentration: Dict[str, Any] = Field(default_factory=dict)
    sector_concentration: Dict[str, Any] = Field(default_factory=dict)
    drawdown: Dict[str, Any] = Field(default_factory=dict)
    stop_loss: Dict[str, Any] = Field(default_factory=dict)


class AllocationTargetCreateRequest(BaseModel):
    account_id: Optional[int] = None
    symbol: Optional[str] = Field(None, max_length=16)
    sector: Optional[str] = Field(None, max_length=32)
    target_pct: float = Field(..., gt=0, le=100)
    drift_threshold_pct: float = Field(2.0, gt=0, le=100)
    priority: int = Field(0, ge=0)


class AllocationTargetUpdateRequest(BaseModel):
    target_pct: Optional[float] = Field(None, gt=0, le=100)
    drift_threshold_pct: Optional[float] = Field(None, gt=0, le=100)
    priority: Optional[int] = Field(None, ge=0)


class AllocationTargetItem(BaseModel):
    id: int
    account_id: Optional[int] = None
    symbol: Optional[str] = None
    sector: Optional[str] = None
    target_pct: float
    drift_threshold_pct: float
    priority: int
    actual_pct: Optional[float] = None
    drift_pct: Optional[float] = None
    is_alert: Optional[bool] = None


class AllocationTargetListResponse(BaseModel):
    targets: List[AllocationTargetItem] = Field(default_factory=list)


class StagedRuleCreateRequest(BaseModel):
    account_id: Optional[int] = None
    symbol: str = Field(..., min_length=1, max_length=16)
    total_target_shares: float = Field(..., gt=0)
    stage_count: int = Field(3, ge=2, le=4)
    stage_pct_1: float = Field(33.33, gt=0, le=100)
    stage_pct_2: Optional[float] = Field(None, gt=0, le=100)
    stage_pct_3: Optional[float] = Field(None, gt=0, le=100)
    stage_pct_4: Optional[float] = Field(None, gt=0, le=100)
    dip_threshold_pct_2: Optional[float] = Field(-5.0)
    dip_threshold_pct_3: Optional[float] = Field(-10.0)
    dip_threshold_pct_4: Optional[float] = Field(None)


class StagedRuleItem(BaseModel):
    id: int
    account_id: Optional[int] = None
    symbol: str
    total_target_shares: float
    stage_count: int
    stage_pct_1: float
    stage_pct_2: Optional[float] = None
    stage_pct_3: Optional[float] = None
    stage_pct_4: Optional[float] = None
    dip_threshold_pct_2: Optional[float] = None
    dip_threshold_pct_3: Optional[float] = None
    dip_threshold_pct_4: Optional[float] = None


class StagedRuleListResponse(BaseModel):
    rules: List[StagedRuleItem] = Field(default_factory=list)


class StagedBuyStageItem(BaseModel):
    stage: int
    qty: float
    price_condition: str
    trigger_price: float


class RebalanceSuggestionItem(BaseModel):
    id: Optional[int] = None
    account_id: Optional[int] = None
    suggestion_type: str
    symbol: Optional[str] = None
    action: Literal["buy", "sell"]
    quantity: float
    estimated_price: float
    estimated_amount: float
    reason: Optional[str] = None
    allocation_before_pct: float
    allocation_after_pct: float


class StagedBuyItem(BaseModel):
    id: int
    account_id: Optional[int] = None
    symbol: str
    current_price: float
    current_shares: float
    total_target_shares: float
    remaining_to_buy: float
    stages: List[StagedBuyStageItem] = Field(default_factory=list)


class CashReserveStatus(BaseModel):
    target_pct: float
    actual_pct: float
    target_amount: float
    actual_amount: float
    shortfall_pct: float
    has_shortfall: bool


class RebalanceReportResponse(BaseModel):
    as_of: str
    account_id: Optional[int] = None
    cost_method: str
    has_targets: bool
    allocation_targets: List[AllocationTargetItem] = Field(default_factory=list)
    drift_alert_count: int
    suggestions: List[RebalanceSuggestionItem] = Field(default_factory=list)
    staged_buys: List[StagedBuyItem] = Field(default_factory=list)
    cash_reserve: CashReserveStatus
